### PR TITLE
CPU profiler fixes

### PR DESCRIPTION
**Pull request by ballard26**
Fixes https://github.com/redpanda-data/redpanda/issues/12901

### CODE REVIEW

File: src/core/cpu_profiler.cc [hunk 1]
```diff
@@ -123,15 +123,31 @@ void cpu_profiler::on_signal() {
...
+    // where the stack is being modified so execution can be returned to
+    // a handler for the exception. This modification isn't capture by 
+    // the eh_frames for the program though. So when libgcc's backtrace
+    // enters the partially modified stack it will follow invalid addresses
+    // and cause a segfault. To avoid this we check if any exception
+    // is currently being unwound and avoid taking a profiling sample if so.
+    //
+    // Note: this only protects against C++ exceptions, therefore foreign
+    // exceptions or long jumps could still cause segfaults within the profiler.
+    const bool no_uncaught_exceptions = std::uncaught_exceptions() == 0;
```
Comment (StephanDollberg) [inline src/core/cpu_profiler.cc:136]: Maybe this is a docs issue but they say:

```
2) Detects how many exceptions in the current thread have been thrown or rethrown and not yet entered their matching catch clauses.
```

So once we enter the catch this should return 0? From my understanding the issue is profiling in the catch handler?

**Review by StephanDollberg — COMMENTED** [review]
I am wondering how much of code actually happens in exception handlers. Good idea with the metrics.

File: src/core/cpu_profiler.cc (hunk 1 above)
Comment (StephanDollberg) [reply src/core/cpu_profiler.cc:136]: Right and so whatever `std::uncaught_exceptions` returns is modified in _cxa_throw and _cxa_catch?

Comment (ballard26) [reply src/core/cpu_profiler.cc:136]: Yeah, sorry should've made that clearer. `std::uncaught_exceptions` just returns the mentioned thread-local counter.

File: tests/unit/cpu_profiler_alloc_test.cc [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+template <typename Func>
+SEASTAR_CONCEPT(requires requires(Func fn) { fn(); })
+void check_function_allocation(const char *name, size_t expected_allocs,
+                               Func f) {
+  auto before = seastar::memory::stats();
+  f();
+  auto after = seastar::memory::stats();
+
+  BOOST_TEST_INFO("After function: " << name);
+  BOOST_REQUIRE_EQUAL(expected_allocs, after.mallocs() - before.mallocs());
```
Comment (StephanDollberg) [inline tests/unit/cpu_profiler_alloc_test.cc:41]: also check before.mallocs() > 0 (will require release mode)? Otherwise we might be not testing anything.

**Review by StephanDollberg — APPROVED** [review]

File: src/core/cpu_profiler.cc (hunk 1 above)
Comment (ballard26) [reply src/core/cpu_profiler.cc:136]: For way too much information see https://refspecs.linuxfoundation.org/abi-eh-1.22.html#cxx-throw

File: tests/unit/cpu_profiler_alloc_test.cc (hunk 2 above)
Comment (ballard26) [reply tests/unit/cpu_profiler_alloc_test.cc:41]: Not sure what we'd gain by checking if `before.mallocs() > 0` as we're only concerned with testing if the `after - before` difference is equal to some amount(0 for this case). A concern could be that the compiler may reorder `f()` outside of the assignments of before and after. In which case we wouldn't be testing anything. Was this your concern?

**Pull request merged**